src/useFetch.ts

```diff
@@ -1,73 +1,130 @@
 import { useEffect, useState, useCallback, useRef, useContext, useMemo } from 'react'
 import FetchContext from './FetchContext'
-import { HTTPMethod, Options, UseFetch, FetchCommands, DestructuringCommands, UseFetchResult, useFetchArg1 } from "./types"
-import { invariant, isObject } from './utils'
-
-
-export function useFetch<TData = any>(arg1?: useFetchArg1, arg2?: Options | RequestInit): UseFetch<TData> {
+import { HTTPMethod, /* Options, */ UseFetch, FetchCommands, DestructuringCommands, UseFetchResult } from './types'
+import { OptionsNoURLs, OptionsAsFirstParam, OptionsAsFirstParamWithContext, URLOrOptions, UseFetchOptions } from './types'
+import { invariant, isObject, isString, pullOutRequestInit } from './utils'
+
+// type Options = OptionsAsFirstParam | OptionsAsFirstParamWithContext | OptionsNoURLs
+
+// No <Provider url='example.com' />
+function useFetch<TData = any>(url: string, options?: OptionsNoURLs): UseFetch<TData>
+function useFetch<TData = any>(options: OptionsAsFirstParam): UseFetch<TData>
+// With <Provider url='example.com' />
+// options should be extended. In future maybe have options callback to completely overwrite options
+// i.e. useFetch('ex.com', oldOptions => ({ ...newOptions })) to overwrite
+function useFetch<TData = any>(url?: string, options?: OptionsNoURLs): UseFetch<TData>
+function useFetch<TData = any>(options?: OptionsAsFirstParamWithContext): UseFetch<TData>
+
+// TODO: handle context.graphql
+function useFetch<TData = any>(urlOrOptions?: URLOrOptions, optionsNoURLs?: OptionsNoURLs): UseFetch<TData> {
   const context = useContext(FetchContext)
 
-  invariant(!!arg1 && !!context.url, 'The first argument of useFetch is required unless you have a global url setup like: <Provider url="https://example.com"></Provider>')
-
-  let url: string | null = context.url || null
-  let options = {} as { signal?: AbortSignal | null } & RequestInit
-  let onMount = false
-  let baseUrl = ''
-  let method: string = HTTPMethod.GET
-
-  const handleOptions = useCallback((opts: Options & RequestInit) => {
-    if (true) {
-      // take out all the things that are not normal `fetch` options
-      // need to take this out of scope so can set the variables below correctly
-      let { url, onMount, timeout, baseUrl, ...rest } = opts
-      options = { signal: undefined, ...rest }
-    }
-    if (context.url) url = context.url
-    if (opts.url) url = opts.url || context.url || ''
-    if (opts.onMount) onMount = opts.onMount
-    if (opts.method) method = opts.method
-    if (opts.baseUrl) baseUrl = opts.baseUrl
+  // TODO: this needs to be per initial setup below since we need to check urlOrOptions.url OR  urlOrOptions.baseUrl
+  invariant(!!urlOrOptions && !!context.url, 'The first argument of useFetch is required unless you have a global url setup like: <Provider url="https://example.com"></Provider>')
```

Comment:
could we somehow have it that `url` is always a string as the first argument?  This would greatly simplify the interface.  If the only time that the url is not a string is when it is being called through the context, could there be a layer of indirection that calls useFetch to keep the interface clean?

Reply:
I like this idea and it does greatly simplify things. AS LONG AS `baseURL` is not necessary. If it is, then we should probably support this. For now, I think we can go with your suggestion and just keep the first  argument as a string and later if we need to, we can edit. On this note I want to mention, @gaearon suggested that we remove the `onMount` option.  After some thought, I actually think this is a good idea because 1, it simplifies things, and 2. it will more-so match the builtin JS `fetch` arguments... meaning it will make this a little more extensible with less effort.
![image](https://user-images.githubusercontent.com/5455859/60383384-afc60080-9a24-11e9-91fb-843dcf6cb00a.png)

Reply:
I'm not saying do a way with baseUrl, just thinking of a way of firming up the arguments so that it is not a case of anything goes.

I think what he is saying is that a blind `useEffect(() => {}, [])` is bad because it is like the old `componentDidMount` but is he not saying there should be another way of triggering the fetch?

It really becomes just a wrapper around `fetch` then and seems much less useful unless I am missing something.  I want loading and error states all handled by the lib

Reply:
> could we somehow have it that `url` is always a string as the first argument? This would greatly simplify the interface. If the only time that the url is not a string is when it is being called through the context, could there be a layer of indirection that calls useFetch to keep the interface clean?

After thinking about this some more, I'm interested. How do you suggest this would look?